pkg/controller/external.go

```diff
@@ -0,0 +1,276 @@
+	"github.com/crossplane-contrib/terrajet/pkg/resource/json"
+	"github.com/crossplane-contrib/terrajet/pkg/terraform"
+)
+
+const (
+	errUnexpectedObject  = "the managed resource is not a Terraformed resource"
+	errGetTerraformSetup = "cannot get terraform setup"
+	errGetWorkspace      = "cannot get a terraform workspace for resource"
+	errRefresh           = "cannot run refresh"
+	errPlan              = "cannot run plan"
+	errStartAsyncApply   = "cannot start async apply"
+	errStartAsyncDestroy = "cannot start async destroy"
+	errApply             = "cannot apply"
+	errDestroy           = "cannot destroy"
+	errStatusUpdate      = "cannot update status of custom resource"
+)
+
+// Option allows you to configure Connector.
+type Option func(*Connector)
+
+// UseAsync configures the controller to use async variant of the functions
+// of the Terraform client.
+func UseAsync() Option {
+	return func(c *Connector) {
+		c.async = true
+	}
+}
+
+// NewConnector returns a new Connector object.
+func NewConnector(kube client.Client, ws Store, sf terraform.SetupFn, opts ...Option) *Connector {
+	c := &Connector{
+		kube:              kube,
+		getTerraformSetup: sf,
+		store:             ws,
+	}
+	for _, f := range opts {
+		f(c)
+	}
+	return c
+}
+
+// Connector initializes the external client with credentials and other configuration
+// parameters.
+type Connector struct {
+	kube              client.Client
+	store             Store
+	getTerraformSetup terraform.SetupFn
+	async             bool
+}
+
+// Connect makes sure the underlying client is ready to issue requests to the
+// provider API.
+func (c *Connector) Connect(ctx context.Context, mg xpresource.Managed) (managed.ExternalClient, error) {
+	tr, ok := mg.(resource.Terraformed)
+	if !ok {
+		return nil, errors.New(errUnexpectedObject)
+	}
+
+	ts, err := c.getTerraformSetup(ctx, c.kube, mg)
+	if err != nil {
+		return nil, errors.Wrap(err, errGetTerraformSetup)
+	}
+
+	tf, err := c.store.Workspace(ctx, tr, ts)
+	if err != nil {
+		return nil, errors.Wrap(err, errGetWorkspace)
+	}
+
+	return &external{
+		kube:      c.kube,
+		workspace: tf,
+		async:     c.async,
+	}, nil
+}
+
+type external struct {
+	kube      client.Client
+	workspace Workspace
+
+	async bool
+}
+
+func (e *external) Observe(ctx context.Context, mg xpresource.Managed) (managed.ExternalObservation, error) { // nolint:gocyclo
+	// We skip the gocyclo check because most of the operations are straight-forward
+	// and serial.
+	// TODO(muvaf): Look for ways to reduce the cyclomatic complexity without
+	// increasing the difficulty of understanding the flow.
+	tr, ok := mg.(resource.Terraformed)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errUnexpectedObject)
+	}
+
+	res, err := e.workspace.Refresh(ctx)
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(err, errRefresh)
+	}
+	if e.async {
+		tr.SetConditions(resource.LastOperationCondition(res.LastOperationError))
```

Comment:
When the creation fails, we store the last error and then doesn't start another operation unless `ApplyAsync` is called. In that case there is no ongoing operation, it goes directly to `!res.Exists` case (in case of creation). So, we'd fail to persist the error if we don't make this call before the `switch` statement.